File: 프로그래머스/2/131127. 할인 행사/할인 행사.js
function solution(want, number, discount) {
    let answer = 0;
    const n = want.length;
    const m = discount.length;

    const wants = {};
    for (let i = 0; i < n; i++) {
        wants[want[i]] = number[i];
    }

    const sales = {};
    for (let i = 0; i < 10; i++) {
        sales[discount[i]] = (sales[discount[i]] || 0) + 1;
    }

    const isPossible = () => {
        for (let name of want) {
            if (!sales[name] || sales[name] < wants[name]) {
                return false;
            }
        }
        return true;
    };

    if (isPossible()) answer++;

    for (let j = 10; j < m; j++) {
        sales[discount[j]] = (sales[discount[j]] || 0) + 1;
        sales[discount[j-10]]--;

        if (sales[discount[j-10]] === 0) {
            delete sales[discount[j-10]];
        }

        if (isPossible()) answer++;
    }

    return answer;
}

File: 프로그래머스/2/131127. 할인 행사/할인 행사.java
import java.util.*;

class Solution {
    static boolean isAllValueZero (String[] want, HashMap<String, Integer> wants) {
        for (String name: want) {
            int val = wants.get(name);
            if (val > 0) return false;
        }
        return true;
    }
    
    public int solution(String[] want, int[] number, String[] discount) {
        int answer = 0;
        HashMap<String, Integer> wants = new HashMap<>();
        for (int i=0; i<want.length; i++) {
            String name = want[i];
            Integer cnt = number[i];
            wants.put(name, cnt);
        }
        // 초기 10일 결과 반영
        for (int idx = 0; idx<10; idx++) {
            String name = discount[idx];
            if (wants.containsKey(name)) {
                int cnt = wants.get(name);
                wants.put(name, cnt - 1);
            }
        }
        // wants가 모두 0인지?
        if (isAllValueZero(want, wants)) answer++;
        
        // 시작일을 늘리며 wants 조작
        for (int st=1; st<=discount.length - 10; st++) {
            int en = st + 10 - 1;
            String subName = discount[st-1];
            String addName = discount[en];
            if (wants.containsKey(subName)) {
                int val = wants.get(subName);
                wants.put(subName, val + 1);
            }
            if (wants.containsKey(addName)) {
                int val = wants.get(addName);
                wants.put(addName, val - 1);
            }
            
            if (isAllValueZero(want, wants)) answer++;
        }
        
        return answer;
    }
}